### DISCUSSION TITLE
Custom syntax highlight on pattern/regex?

### DISCUSSION THREAD

**Discussion by johnalotoski** [akinomyoga/ble.sh#362] [Q&A] [answered]
Is there a way to syntax highlight a user defined pattern in a special way?

Example: Custom highlight regex `^rm -rf.*` with bright red background, white foreground in ble.sh?

Similar example in zsh, [link](https://github.com/disassembler/network/blob/master/modules/profiles/zsh.nix#L26-L31)

**Comment by akinomyoga** [comment, marked as answer]
I'm sorry for my late reply. I missed your post since I was busy that week.

There is currently no easy interface to configure highlighting patterns. In principle, you could implement a custom *layer* as shell functions (`ble/highlight/layer:<LAYER>/*` with `<LAYER>` being the name of the new layer). The details of implementing a layer is described in [a code comment](https://github.com/akinomyoga/ble.sh/blob/bdcecbbff25db9bc17e081b691c899870b0fd4a5/src/color.sh#L1524-L1593), but it's written in Japanese (so please use machine translations if you are really interested). I think [the implementation of the `region` layer](https://github.com/akinomyoga/ble.sh/blob/bdcecbbff25db9bc17e081b691c899870b0fd4a5/src/color.sh#L1676-L1849) can be referenced as an example.

Maybe, when I have time, I can implement a new layer that offers a simple interface for configuring highlighting by patterns.

**Reply by johnalotoski** [reply]
> I'm sorry for my late reply. I missed your post since I was busy that week.
There is currently no easy interface to configure highlighting patterns. In principle, you could implement a custom layer as shell functions (ble/highlight/layer:<LAYER>/* with <LAYER> being the name of the new layer). The details of implementing a layer is described in a code comment, but it's written in Japanese (so please use machine translations if you are really interested). I think the implementation of the region layer can be referenced as an example.

Thank you for the information, I'll look into that.

> Maybe, when I have time, I can implement a new layer that offers a simple interface for configuring highlighting by patterns.

That would be awesome!

**Comment by akinomyoga** [comment]
> Maybe, when I have time, I can implement a new layer that offers a simple interface for configuring highlighting by patterns.

I have implemented a new layer class in [`contrib/layer/pattern.bash`](https://github.com/akinomyoga/blesh-contrib/blob/add8d64cd66c1d51437b1a009d31b6602d10805c/layer/pattern.bash). You can update ble.sh by running `ble-update` in a ble session, and e.g. write the following settings in blerc:

```bash
ble-import layer/pattern

# Create a layer with the name "pattern1".
# The patterns in this layer are by default interpreted as regular expressions.
ble/highlight/layer:{pattern}/declare pattern1
ble/highlight/layer:{pattern}/register pattern1 'rm -rf [^;&|]*' 'fg=white,bold,bg=red'    # add a pattern in a regular expression
ble/array#insert-after _ble_highlight_layer_list syntax pattern1    # register the layer "pattern1" to the layer list.

# Create another layer with the name "pattern2".
# The argument "glob" specifies that the patterns in this layer should be treated as glob patterns.
ble/highlight/layer:{pattern}/declare pattern2 glob
ble/highlight/layer:{pattern}/register pattern2 "$USER" 'fg=blue,bold'
ble/highlight/layer:{pattern}/register pattern2 "$HOSTNAME" 'fg=green,bold'    # one can add multiple patterns in a pattern layer
ble/highlight/layer:{pattern}/register pattern2 '[0-9]+' 'bg=216,fg=black'
ble/array#insert-after _ble_highlight_layer_list pattern1 pattern2
```

Detailed descriptions on the functions can be found in the beginning comment of [`contrib/layer/pattern.bash`](https://github.com/akinomyoga/blesh-contrib/blob/add8d64cd66c1d51437b1a009d31b6602d10805c/layer/pattern.bash).

**Reply by johnalotoski** [reply]
Awesome, thank you very much!
This is great!
